MREUnityRuntime/MREUnityRuntimeLib/Assets/AssetCache.cs

```diff
@@ -59,20 +62,41 @@ public GameObject EmptyTemplate()
         /// <inheritdoc cref="GetAsset"/>
         public Object GetAsset(Guid? id)
         {
-            return cache.Find(c => c.Id == id).Asset;
+            return id != null ? cache.Find(c => c.Id == id).Asset : null;
         }
 
         /// <inheritdoc cref="GetId"/>
         public Guid? GetId(Object asset)
         {
-            var id = cache.Find(c => c.Asset == asset).Id;
-            return id != Guid.Empty ? (Guid?)id : null;
+            return asset != null ? cache.Find(c => c.Asset == asset).Id : (Guid?)null;
+        }
+
+        /// <inheritdoc cref="OnCached"/>
+        public void OnCached(Guid id, CacheCallback callback)
+        {
+            var asset = GetAsset(id);
+            if (asset != null)
+            {
+                callback?.Invoke(asset);
+            }
+            else
+            {
+                cacheCallbacks.GetOrCreate(id, () => new List<CacheCallback>(3)).Add(callback);
+            }
```

Comment:
I'd like if you could only call OnCached on assets that are actively being cached - currently you can use use OnCached on any unloaded asset ID, and it will just wait forever. Could we change CacheAsset to CacheAssetStart and CacheAssetComplete, and CacheAssetStart adds an entry so you're allowed to do OnCached on it, and otherwise it fails?

Reply:
Similarly, if asset loading fails/times out, we should have a way to deal with this. CacheCallback may be called with null on failure?


---
In reply to: [306969021](https://github.com/Microsoft/mixed-reality-extension-unity/pull/92#discussion_r306969021) [](ancestors = 306969021)

Reply:
Yes, waiting for an ID that will not be loaded would leak a little memory and cause the callback to never be executed. I don't think that can happen this far down the stack though, we have lots of checks in the SDK that the asset exists before we assign it. And unloaded assets don't exist for this purpose. If the user insists on doing something dumb, like generating a new guid and assigning that, it becomes `null` long before it hits Unity.

---
In reply to: [306969021](https://github.com/microsoft/mixed-reality-extension-unity/pull/92#discussion_r306969021) [](ancestors = 306969021)

Reply:
Resolved by caching a `null` value when the asset fails to load. Permanently marks that guid as a dud.

---
In reply to: [306973398](https://github.com/microsoft/mixed-reality-extension-unity/pull/92#discussion_r306973398) [](ancestors = 306973398,306969021)